### ISSUE TITLE
Make WC_Logger extensible

### ISSUE THREAD

**Issue by gedex** [woocommerce/woocommerce#10468]
Currently `WC_Logger` only writes to a file system. It would be great to make it extensible by allowing handlers and formatters to be swapped.

Handlers are implementing common methods like opening and writing records. The target could be file system, databases, or services (like logstash).

Formatters are implementing method that returns how record should be written, it could be plain text line or JSON.
## Why
- Reading/writing into file system is expensive operation when file size is big
- It uses upload dir which is publicly accessible by default
- For big merchants that log every transaction it would be useful to store log records in dedicated logging system like database or logstash so that it can be queried when needed.

**Comment by roykho** [comment]
:heart:

**Comment by claudiosmweb** [comment]
We should have at least a rotation system.

**Comment by dnaber-de** [comment]
It seems that [my question on the related PR](https://github.com/woocommerce/woocommerce/pull/12340#issuecomment-262925691) was deleted. Maybe it was the wrong place to ask, so sorry for the confusion. However, I wonder why the patch »loosely adhere to the PHP Standards Recommendations« instead of implementing them. The benefit would be that one could replace the logger and use for example Monolog, which has a large set of very useful handlers.

**Comment by sirreal** [comment]
@dnaber-de I'm the PR author and I'm happy to receive your input, I think the question is a valid one.

I decided to "loosely" adhere rather than "strictly" adhere for 2 reasons:

1. `WC_Logger` currently accepts all logs through the `add` method. This is not part of the spec and I wanted to maintain it for backwards compatibility. I *did* deprecate this method.
2. The following part of the spec did not seem appropriate to me in this context. Rather than throw an exception, I added a `_doing_it_wrong` message.
    > Calling this method with a level not defined by this specification MUST throw a `Psr\Log\InvalidArgumentException` if the implementation does not know about the level

The rest of the spec essentially describes the nine methods:

```php
log( $level, $message, array $context = array() );
emergency( $message, array $context = array() );
// ...
debug( $message, array $context = array() );
```

This part of the spec *is* strictly respected, and I assume it's what logger implementors are expecting.
